### PR TITLE
Integrate runtime components into single runtime

### PR DESCRIPTION
**Pull request by fw-immunant**
This is not yet applied to our compartmentalized binaries and tests, but runs a tiny (uncompartmentalized) demo program I've been linking against it manually for validation.

This PR will be ready for merging when the build-system work is done to integrate the runtime with our tests themselves, but we'll want a switch to build without the runtime (or disable it) so that we can debug tests, as the runtime (being a `ptrace`r) conflicts with gdb and other debuggers.

This implements an initial coarse-grained syscall policy very similar to that described in #233, and integrates fine-grained syscall filtering as described in #235. As we validate this on our tests/demons/nginx we'll probably add syscalls to our allowlist to get closer to what #231 outlines.

### CODE REVIEW

**Comment by fw-immunant** [comment]
One wart here is that we use `SECCOMP_FILTER_FLAG_NEW_LISTENER` to get a seccomp_unotify fd, but we don't actually use this fd for anything. If we need to do more detailed inspection of syscall arguments in memory or if we need to inject fds into the inferior to simulate some syscalls, we likely will need this fd; see `man 2 seccomp_unotify`. But for now, we're setting up this infrastructure without it being load-bearing.

File: runtime/seccomp_filter.c [hunk 1]
```diff
@@ -79,6 +70,6 @@ int configure_seccomp(void) {
     perror("seccomp(SECCOMP_FILTER_FLAG_TSYNC)");
     return -1;
   }
-  // prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER, &prog);
```
Comment (fw-immunant) [inline runtime/seccomp_filter.c:85]: This was debugging cruft; this form is deprecated and strictly less flexible compared to the standalone `seccomp()` syscall.

File: runtime/seccomp_filter_demo.c [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+#include "seccomp_filter.h"
+
+int main(int argc, char *argv[]) {
+  int infd, outfd;
+  ssize_t read_bytes;
+  char buffer[1024];
+
+  if (argc < 3) {
+    printf("Usage:\n\tdup_file <input path> <output_path>\n");
```
Comment (rinon) [inline runtime/seccomp_filter_demo.c:15]: dup_file?

File: runtime/track_memory_map.c [hunk 3]
```diff
@@ -346,40 +359,51 @@ void return_syscall_eperm(pid_t pid) {
   fprintf(stderr, "wrote -eperm to rax\n");
 }
 
-void track_memory_map(pid_t pid, struct memory_map *map) {
+/* track the inferior process' memory map.
+
+returns true if the inferior exits, false on trace error.
+
+if true is returned, the inferior's exit status will be stored to *exit_status_out if not NULL. */
```
Comment (rinon) [inline runtime/track_memory_map.c:366]: Probably want to put this in the header?

**Review by rinon — COMMENTED** [review]
So as is the usage for this is to add the `ia2-main` library which wraps main to fork and run the binary? How does this interact with our existing wrapping of main to set the pkey?

Did you decide to do this rather than the external binary that runs a separate child with fork+exec? Or was this implementation before we discussed that?

File: runtime/CMakeLists.txt [hunk 4]
```diff
@@ -26,11 +26,26 @@ add_executable(track-memory-map
...
 )
 
 add_executable(landlock
+  landlock_demo.c
   landlock.c
   strv.c
   forbid_paths.c
 )
+
+add_library(ia2-main
```
[the remark this answers is not in the document]
Comment (fw-immunant) [reply runtime/CMakeLists.txt]: Now called `ia2-sandbox`.

**Comment by fw-immunant** [comment]
This is without the changes needed to support multiple threads and fork/exec; see #307.

**Pull request merged**